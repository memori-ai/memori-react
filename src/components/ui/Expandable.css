.memori-expandable {
    line-height: 1;
}

.memori-expandable--inner {
    overflow: hidden;
    max-height: 9999px;
    transition: max-height 0.2s ease-in-out;
}

.memori-expandable>button {
    color: var(--memori-primary, #333);
    font-size: 0.9em;
    text-decoration: underline
}
