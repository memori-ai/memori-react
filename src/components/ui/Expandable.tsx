import React, { useEffect, useRef, useState } from 'react';
import Button from './Button';
import cx from 'classnames';
import { useTranslation } from 'react-i18next';

export interface Props {
  rows: number;
  className?: string;
  innerClassName?: string;
  btnClassName?: string;
  lineHeightMultiplier?: number;
  defaultExpanded?: boolean;
  expandSymbol?: (lang: string) => React.ReactNode;
  collapseSymbol?: (lang: string) => React.ReactNode;
  children: React.ReactNode;
}

const Expandable = ({
  rows,
  className,
  innerClassName,
  btnClassName,
  lineHeightMultiplier = 1.2,
  defaultExpanded = false,
  expandSymbol = () => '...',
  collapseSymbol = (lang: string) =>
    lang === 'it' ? 'Mostra meno' : 'Show less',
  children,
}: Props) => {
  const { i18n } = useTranslation();
  const lang = i18n.language;
  const collapseSymbolText = collapseSymbol(lang);
  const expandSymbolText = expandSymbol(lang);
  const [expanded, setExpanded] = useState(defaultExpanded);
  const [needsExpanding, setNeedsExpanding] = useState(false);
  const [rowHeight, setRowHeight] = useState(16);
  const ref = useRef(null);

  useEffect(() => {
    if (ref.current) {
      let height = (ref.current as HTMLElement).getBoundingClientRect().height;
      let computedStyle = getComputedStyle(ref.current as HTMLElement);
      let elLineHeight = computedStyle.lineHeight;
      let lineHeight =
        elLineHeight === 'normal' || !elLineHeight?.length
          ? lineHeightMultiplier * parseInt(computedStyle.fontSize, 10)
          : parseInt(elLineHeight, 10);

      console.table({
        rows,
        lineHeight,
        height,
        maxHeight: rows * lineHeight,
        needsExpanding: height > rows * lineHeight,
      });
      setRowHeight(lineHeight);
      if (height && height > rows * lineHeight) {
        setNeedsExpanding(true);
      }
    }
  }, [rows, ref.current]);

  return (
    <div className={cx('memori-expandable', className)}>
      <div
        ref={ref}
        className={cx('memori-expandable--inner', innerClassName)}
        style={{
          maxHeight:
            expanded || !needsExpanding ? '9999px' : `${rowHeight * rows}px`,
        }}
      >
        {children}
      </div>
      {needsExpanding && !expanded && (
        <Button
          ghost
          padded={false}
          className={btnClassName}
          onClick={() => setExpanded(true)}
        >
          {expandSymbolText}
        </Button>
      )}
      {needsExpanding && expanded && (
        <Button
          ghost
          padded={false}
          className={btnClassName}
          onClick={() => setExpanded(false)}
        >
          {collapseSymbolText}
        </Button>
      )}
    </div>
  );
};

export default Expandable;
