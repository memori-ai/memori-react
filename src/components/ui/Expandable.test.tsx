import { render } from '@testing-library/react';
import Expandable from './Expandable';

it('renders Expandable unchanged', () => {
  const { container } = render(<Expandable rows={1}>Test</Expandable>);
  expect(container).toMatchSnapshot();
});

it('renders Expandable with long text unchanged', () => {
  const { container } = render(
    <Expandable rows={1}>
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla,
      sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse
      a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper
      nisi.
    </Expandable>
  );
  expect(container).toMatchSnapshot();
});

it('renders Expandable with multiple rows unchanged', () => {
  const { container } = render(
    <Expandable rows={3}>
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla,
      sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse
      a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper
      nisi.
    </Expandable>
  );
  expect(container).toMatchSnapshot();
});

it('renders Expandable with custom wrapper CSS class unchanged', () => {
  const { container } = render(
    <Expandable rows={1} className="test">
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla,
      sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse
      a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper
      nisi.
    </Expandable>
  );
  expect(container).toMatchSnapshot();
});

it('renders Expandable with custom inner CSS class unchanged', () => {
  const { container } = render(
    <Expandable rows={1} innerClassName="test">
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla,
      sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse
      a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper
      nisi.
    </Expandable>
  );
  expect(container).toMatchSnapshot();
});

it('renders Expandable with custom button CSS class unchanged', () => {
  const { container } = render(
    <Expandable rows={1} btnClassName="test">
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla,
      sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse
      a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper
      nisi.
    </Expandable>
  );
  expect(container).toMatchSnapshot();
});

it('renders Expandable with custom expand symbol unchanged', () => {
  const { container } = render(
    <Expandable rows={1} expandSymbol={() => 'test'}>
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla,
      sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse
      a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper
      nisi.
    </Expandable>
  );
  expect(container).toMatchSnapshot();
});

it('renders Expandable with custom collapse symbol unchanged', () => {
  const { container } = render(
    <Expandable rows={1} collapseSymbol={() => 'test collapse'} defaultExpanded>
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla,
      sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse
      a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper
      nisi.
    </Expandable>
  );
  expect(container).toMatchSnapshot();
});

it('renders Expandable with JSX markup unchanged', () => {
  const { container } = render(
    <Expandable rows={1}>
      <h1>Lorem ipsum</h1>
      <p>Suspendisse a sodales nulla, sed semper nisi.</p>
      <h2>Suspendisse a sodales nulla, sed semper nisi.</h2>
      <p>Suspendisse a sodales nulla, sed semper nisi.</p>
      <p>Suspendisse a sodales nulla, sed semper nisi.</p>
      <button>Dolor sit amet</button>
      <p>Suspendisse a sodales nulla, sed semper nisi.</p>
      <p>Suspendisse a sodales nulla, sed semper nisi.</p>
    </Expandable>
  );
  expect(container).toMatchSnapshot();
});
