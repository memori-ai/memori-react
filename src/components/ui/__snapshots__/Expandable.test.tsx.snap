// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Expandable unchanged 1`] = `
<div>
  <div
    class="memori-expandable"
  >
    <div
      class="memori-expandable--inner"
      style="max-height: 9999px;"
    >
      Test
    </div>
  </div>
</div>
`;

exports[`renders Expandable with JSX markup unchanged 1`] = `
<div>
  <div
    class="memori-expandable"
  >
    <div
      class="memori-expandable--inner"
      style="max-height: 9999px;"
    >
      <h1>
        Lorem ipsum
      </h1>
      <p>
        Suspendisse a sodales nulla, sed semper nisi.
      </p>
      <h2>
        Suspendisse a sodales nulla, sed semper nisi.
      </h2>
      <p>
        Suspendisse a sodales nulla, sed semper nisi.
      </p>
      <p>
        Suspendisse a sodales nulla, sed semper nisi.
      </p>
      <button>
        Dolor sit amet
      </button>
      <p>
        Suspendisse a sodales nulla, sed semper nisi.
      </p>
      <p>
        Suspendisse a sodales nulla, sed semper nisi.
      </p>
    </div>
  </div>
</div>
`;

exports[`renders Expandable with custom button CSS class unchanged 1`] = `
<div>
  <div
    class="memori-expandable"
  >
    <div
      class="memori-expandable--inner"
      style="max-height: 9999px;"
    >
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi.
    </div>
  </div>
</div>
`;

exports[`renders Expandable with custom collapse symbol unchanged 1`] = `
<div>
  <div
    class="memori-expandable"
  >
    <div
      class="memori-expandable--inner"
      style="max-height: 9999px;"
    >
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi.
    </div>
  </div>
</div>
`;

exports[`renders Expandable with custom expand symbol unchanged 1`] = `
<div>
  <div
    class="memori-expandable"
  >
    <div
      class="memori-expandable--inner"
      style="max-height: 9999px;"
    >
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi.
    </div>
  </div>
</div>
`;

exports[`renders Expandable with custom inner CSS class unchanged 1`] = `
<div>
  <div
    class="memori-expandable"
  >
    <div
      class="memori-expandable--inner test"
      style="max-height: 9999px;"
    >
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi.
    </div>
  </div>
</div>
`;

exports[`renders Expandable with custom wrapper CSS class unchanged 1`] = `
<div>
  <div
    class="memori-expandable test"
  >
    <div
      class="memori-expandable--inner"
      style="max-height: 9999px;"
    >
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi.
    </div>
  </div>
</div>
`;

exports[`renders Expandable with long text unchanged 1`] = `
<div>
  <div
    class="memori-expandable"
  >
    <div
      class="memori-expandable--inner"
      style="max-height: 9999px;"
    >
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi.
    </div>
  </div>
</div>
`;

exports[`renders Expandable with multiple rows unchanged 1`] = `
<div>
  <div
    class="memori-expandable"
  >
    <div
      class="memori-expandable--inner"
      style="max-height: 9999px;"
    >
      Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi. Suspendisse a sodales nulla, sed semper nisi.
    </div>
  </div>
</div>
`;
